bugsnag-android-core/src/main/java/com/bugsnag/android/internal/JournaledDocument.kt

```diff
@@ -0,0 +1,171 @@
+     */
+    fun close() {
+        journalStream.close()
+        isOpen = false
+    }
+
+    override val entries: Set<Map.Entry<String, Any>> get() = document.entries
+    override val keys: Set<String> get() = document.keys
+    override val size: Int get() = document.size
+    override val values: Collection<Any> get() = document.values
+    override fun containsKey(key: String): Boolean { return document.containsKey(key) }
+    override fun containsValue(value: Any): Boolean { return document.containsValue(value) }
+    @RequiresApi(Build.VERSION_CODES.N)
+    override fun forEach(action: BiConsumer<in String, in Any>) { document.forEach(action) }
+    override fun get(key: String): Any? { return document[key] }
+    @RequiresApi(Build.VERSION_CODES.N)
+    override fun getOrDefault(key: String, defaultValue: Any): Any { return document.getOrDefault(key, defaultValue) }
+    override fun isEmpty(): Boolean { return document.isEmpty() }
+
+    companion object {
+        // 0x99 is guaranteed to be an invalid UTF-8 start byte
+        const val clearedByteValue: Byte = 0x99.toByte()
+
+        private const val snapshotExtension = ".snapshot"
+        private const val journalExtension = ".journal"
+        private const val newSnapshotExtension = ".snapshot.new"
+
+        /**
+         * Check if a journal-backed document exists at this base path.
+         *
+         * @param baseDocumentPath The base path + filename to derive snapshot and journal names from.
+         * @return true if document files exist at this base path.
+         */
+        fun documentExists(baseDocumentPath: File): Boolean {
+            val snapshotPath = File(baseDocumentPath.path + snapshotExtension)
+            val newSnapshotPath = File(baseDocumentPath.path + newSnapshotExtension)
+
+            return newSnapshotPath.exists() || snapshotPath.exists()
+        }
+
+        /**
+         * Return the equivalent MutableMap<String, Any> that matches the document on disk with the
+         * journal on disk applied.
+         *
+         * Because of the many states the files can be in after a crash, this function checks the
+         * files in a very specific way:
+         * - Check for a ".snapshot.new" file and just return that if it's valid.
+         * - Otherwise load the ".snapshot" file, which must be valid.
+         *   - Attempt to load the ".journal" file, and apply it if it's valid.
+         *   - Return the finished document.
+         *
+         * @param baseDocumentPath The base path + filename to derive snapshot and journal names from.
+         * @return The document.
+         */
+        fun loadDocumentContents(baseDocumentPath: File): MutableMap<in String, out Any> {
+            val journalPath = File(baseDocumentPath.path + journalExtension)
+            val snapshotPath = File(baseDocumentPath.path + snapshotExtension)
+            val newSnapshotPath = File(baseDocumentPath.path + newSnapshotExtension)
```

Comment:
nit: there's similar code to calculate these paths in the constructor. It might be worth thinking about some abstraction that makes sure the path is only calculated in one place, in case it changes in future.